### PR TITLE
Speeds up the preference menu, significantly. Adds object pooling, other stuff too

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Hello again bestie.
I'm very miffed about how slow the preferences menu is. Especially in light of #63118
It's very slow, and because it runs off Topic (which is technically a verb) it runs after even sendmaps. Lots of lagspikes.

Sooooo I'm going to optimize it. Need more user experience time for atmos after all.
Let's break down where it slows down/speeds up yeah?

Oh and before I start, all profiles are setup like this.
I intercept the ui_act args intended for preferences, and store them in a list on SSInput. Every time SSInput fires, we call the ui_act once. Preferences handily doesn't cache ui acts, so this is very effective.
Anyway I call the ui_act 1.2k times, and then stop. Hopefully this is a large enough call count to make any issues very clear from the profiler.
 
 Using my ethereal static for this, if it matters to you
image

To start with, here's a profile of updating the preferences as it is on master, no outfit, just bare naked
[Pre Optimization No Gear](https://github.com/tgstation/tgstation/files/7647114/pre.optimization.no.gear.csv)
Takes roughly 30 seconds to run 1.2k calls. Not great, roughly 25 milliseconds

Ok so how long does it take to do the same thing, but to a character wearing a CEs outfit?
[Pre Optimization CE](https://github.com/tgstation/tgstation/files/7647132/pre.optimization.CE.csv)
141 SECONDS. Jesus christ, that's roughly 117 miliseconds, almost a 10th of a second. No fucking wonder this lags sometimes.

Ok so why then. Well the answer is a bit longwinded, but let's start with the start yeah?
Of that 141 seconds, we spend nearly 66 seconds just generating typecaches.
Turns out the storage component is monumentally stupid, and uses string lists to cache a typecache of the list passed into it, so we save on memory by only storing one list, but we lose on cpu time because we need to generate a new typecache for each list, and god trust me these lists hardly vary.

 So anyway that's like 50% of our worst case gone https://github.com/LemonInTheDark/tgstation/commit/9e197b584e6a2631cf7f55eb97309adca8d00ceb
Moving on.

The main reason we need to do so much tine on preferences is because every change to the prefs menu triggers a full rebuild of the dummy.
This is to make preference options development as simple as possible, but it places some pretty significant overhead on reloading, since it turns out spawning in humans is slow.

But why exactly is it slow?
Well there's 2 main painpoints. Painpoint 1 is generating organs. Each time you reload the prefs menu, we need to regenerate your species from a template. This means deleting and respawning your organs.

Painpoint 2 is equipping the mob. Typecaches aside, generating the equipment needed to kit out any reasonably advanced job like paramedic or engineer takes a pretty significant chunk of time.

Starting with organs, we only really need to spawn in organs that you can like, see, right? Like it's not like the dummies breath, or have blood or eat food, so we can just like, not give them stomachs and lungs and such.

That's what https://github.com/LemonInTheDark/tgstation/commit/b78352728a66448d7fcc5480c05d572f24402d26 did.
It didn't save as much time as I claimed in that commit desc, but it's still not a bad amount. 9 organs to ~3 is pretty damn good.

Ok so items. It's not like we can just not spawn them in, since we have no idea what humans need to see and what they don't, and modifying job code to suit that goal is not reasonable.

So instead, why not just, not spawn the items immediately? We could queue them up in a subsystem, and then just hand then out on request.
https://github.com/LemonInTheDark/tgstation/commit/a62b591d5b0f1ee172675571ec54c852b36ef62e

There's a whole lot more to this then just this commit, most of this pr is setup to make this system of object pooling as not bad as possible.
It's by design always going to be a bit fragile, since objects can do strange things on initialize, or have weird expectations.
But if it's used in a limited scope that's not the worst thing to deal with, and god it's so fast.

Ok so things are better, but we still spend a lot of time creating and deleting organs. Let's just queue those too
https://github.com/LemonInTheDark/tgstation/commit/f9d82df3454f702fac6d4887b8130c79550f214f
(I later walked back part of this, queuing limbs is kinda pointless and brains are just more trouble then they're worth)

That's even better, but our cache can fail if someone spams the menu, and we're still spending way more time then we should be deleting equipment and organs.
So why not just like, not? It's not like dummies are running around the map using their equipment, so when we delete it couldn't we insert it back into our queue? Same thing with the organs really, except they need special handling for extraction and insertion to disable/enable their rotting mechanic.

https://github.com/LemonInTheDark/tgstation/commit/c862a783899e3619ce58d16d16d13688eb2cefb2
This is really really fast. We obviously can't do this with everything, if we reinserted backpacks for instance we could end up duping their contents, which is A:weird for players, and B:a potential memory leak.
So let's pretend any of this is sane and just cache the things with no contents. If we want to speed up their generation, we can just cache the contents of the object, so loading it is faster.

Since we're doing this, let's cache the contents of belts, and also pdas. Unshits pda code slightly in the process.

Moving on, things are still a bit slow. In fact somehow ungeared runs are faster without caching then with it.
That's really weird. Most of the extra cost is in regenerate organs and harvest organs, and...
Oh.

Whenever you remove or insert an eye into someone, it calls regenerate_icons on the mob.
I can't even begin to tell you how slow this is, but to put it into perspective it takes roughly 10 seconds of the 40 we're spending loading cached ces.

Allllright, so we can actually just not, since like 4 years ago timegreen added the right proc, update_body (which updates eye color) to the bottom of the proc. 

https://github.com/LemonInTheDark/tgstation/commit/3173bca60d4af7c6308d90db008f78f140d378fc

Well that's much better, ahhhhhhh ok just about done then eh?
Wait a moment, why does my mob have no well, anything?
qc9ciZGV6k

Turns out that the regenerate_icons call in EYE REMOVAL IS THE ONLY REASON THAT DUMMIES RENDER LIMBS RIGHT NOW
https://github.com/LemonInTheDark/tgstation/commit/d817ef053875e38c9ec3552c3eea1856984298da

`update_body_parts` caches its overlay updates based on the makeup of its limbs, but since we empty the mob's overlay list as a part of dummy creation, this does nothing.

dependency

Alllright, finally done.

If you're curious, here are the final profiles
[Cached CE](https://github.com/tgstation/tgstation/files/7647247/cached.CE.csv)
[Cached, No Gear](https://github.com/tgstation/tgstation/files/7647248/cached.no.gear.csv)
[Uncached CE](https://github.com/tgstation/tgstation/files/7647249/uncached.CE.csv)
[Uncached, No Gear](https://github.com/tgstation/tgstation/files/7647250/uncached.no.gear.csv)

Caching saves 1.5x for renders with no gear, and 2x for renders with max gear. So pretty fucking good.

Looking at it from a before and after point of view, No gear rendering has gotten 3 times better, and Fully geared rendering has gotten 7 times faster.

Boooo yah suck it playerbase.

## Why It's Good For The Game

Papa needs some more execution time, OHHH fuck this whole thing just effects lag spikes doesn't it DAMN IT.
IDK maybe it makes the profiler easier?
Oh I guess you'll see less or hopefully no lag spikes at highpop when someone changes the color of their hair from #c034eb to #d934eb

If the lag spikes continue we can just move rerendering to a subsystem, but I'm not totally sure that's needed. We'll have to see.

One other thing, round start code makes significant use of preference dummies. Need to generate profiles for people's medical/security records and all.
I don't really have a good way of telling you how much what I'm doing here helps with that, since `get_flat_human_icon` also well, calls getFlatIcon, but it ought to at least help a bit.

Ah WAIT ONE MORE THING
Since object pooling and the optimizations done to human loading are generic, this speeds up at least somewhat the things that load humans. So roundstart again, just spawning in, etc.

TODO/Think about:
- [ ] Scale the amount of gear we preload to match pop (This is a maybe) 
- [x] Make a unit test that checks to make sure that everything loads proper, and that nothing qdel's itself or something. This way we can at least pretend that the system is robust 

TL;DR
Dummy go speed
Objects are cached and recached sometimes (sus)
Roundstart is maybe faster

## Changelog

:cl:
code: I've changed how job equipment and organs are loaded, if you end up with someone else's head/id card/eyes, please let me know yeah?
code: The preferences menu is now much more efficient. Hopefully you'll see less small jitters at highpop. 
/:cl:

### CODE REVIEW

**Comment by Ghommie** [comment]
> Turns out that the regenerate_icons call in EYE REMOVAL IS THE ONLY REASON THAT DUMMIES RENDER LIMBS RIGHT NOW

Eh, I told mothblocks about this issue roughly 3 months ago on Discord. I pity you too had to go through the pain of searching for the source of this crap and I'm also sorry I didn't PR a fix to that myself. Good job, by the by.

**Comment by CRITAWAKETS** [comment]
lemon dives into species code: becomes insane as a result

**Comment by LemonInTheDark** [comment]
> Eh, I told mothblocks about this issue roughly 3 months ago on Discord. I pity you too had to go through the pain of searching for the source of this crap and I'm also sorry I didn't PR a fix to that myself. Good job, by the by.

It wasn't too bad, but I think that was only because I noticed it right after I changed eyes. if it had been any more drifted it would have been suffering.
Thanks, hopefully this helps things a bit

**Comment by Mothblocks** [comment]
> Eh, I told mothblocks about this issue roughly 3 months ago on Discord

I don't remember this at all LOL

**Comment by LemonInTheDark** [comment]
image
Headcoder, your maintainer is no longer fit for purpose

File: code/controllers/subsystem/wardrobe.dm [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+				canonize_type(species_request)
+		CHECK_TICK
+
+/datum/controller/subsystem/wardrobe/proc/load_pda_nicknacks()
+	for(var/obj/item/pda/pager as anything in typesof(/obj/item/pda))
+		var/obj/item/pda/flip_phone = new pager()
+		for(var/datum/outfit_item_type as anything in flip_phone.get_types_to_preload())
+			canonize_type(outfit_item_type)
+		qdel(flip_phone)
+		CHECK_TICK
```
Comment (Watermelon914) [inline code/controllers/subsystem/wardrobe.dm:332]: these get called in initialize, why use CHECK_TICK?

**Review by Watermelon914 — APPROVED** [review]

File: code/modules/mob/living/carbon/human/dummy.dm [hunk 2]
```diff
@@ -17,6 +18,64 @@ INITIALIZE_IMMEDIATE(/mob/living/carbon/human/dummy)
...
+	harvest_organs()
+	return ..()
+
+///Let's extract our dummies organs and limbs for storage, to reduce the cache missed that spamming a dummy cause
+/mob/living/carbon/human/dummy/proc/harvest_organs()
+	for(var/slot in list(ORGAN_SLOT_BRAIN, ORGAN_SLOT_HEART, ORGAN_SLOT_LUNGS, ORGAN_SLOT_APPENDIX, \
+		ORGAN_SLOT_EYES, ORGAN_SLOT_EARS, ORGAN_SLOT_TONGUE, ORGAN_SLOT_LIVER, ORGAN_SLOT_STOMACH))
+		var/obj/item/organ/current_organ = getorganslot(slot) //Time to cache it lads
+		if(current_organ)
+			current_organ.Remove(src, special=TRUE) //Please don't somehow kill our dummy
```
Comment (SuperNovaa41) [inline code/modules/mob/living/carbon/human/dummy.dm:31]:
```suggestion
			current_organ.Remove(src, special = TRUE) //Please don't somehow kill our dummy
```
https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#spacing

File: code/modules/mob/living/carbon/human/dummy.dm [hunk 3]
```diff
@@ -17,6 +18,64 @@ INITIALIZE_IMMEDIATE(/mob/living/carbon/human/dummy)
...
+			SSwardrobe.stash_object(current_organ)
+
+	var/datum/species/current_species = dna.species
+	for(var/organ_path in current_species.mutant_organs)
+		var/obj/item/organ/current_organ = getorgan(organ_path)
+		if(current_organ)
+			current_organ.Remove(src, special=TRUE) //Please don't somehow kill our dummy
```
Comment (SuperNovaa41) [inline code/modules/mob/living/carbon/human/dummy.dm:38]:
```suggestion
			current_organ.Remove(src, special = TRUE) //Please don't somehow kill our dummy
```

File: code/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -313,8 +316,9 @@ GLOBAL_LIST_EMPTY(features_by_species)
  * * old_species - datum, used when regenerate organs is called in a switching species to remove old mutant organs.
  * * replace_current - boolean, forces all old organs to get deleted whether or not they pass the species' ability to keep that organ
  * * excluded_zones - list, add zone defines to block organs inside of the zones from getting handled. see headless mutation for an example
+ * * visual_only - boolean, only load organs that change how the species looks. Do not use for normal gameplay stuff
  */
-/datum/species/proc/regenerate_organs(mob/living/carbon/C,datum/species/old_species,replace_current=TRUE,list/excluded_zones)
+/datum/species/proc/regenerate_organs(mob/living/carbon/C, datum/species/old_species, replace_current=TRUE, list/excluded_zones, visual_only=FALSE)
```
Comment (SuperNovaa41) [inline code/modules/mob/living/carbon/human/species.dm:321]:
```suggestion
/datum/species/proc/regenerate_organs(mob/living/carbon/C, datum/species/old_species, replace_current = TRUE, list/excluded_zones, visual_only = FALSE)
```

File: code/modules/mob/living/carbon/human/species_types/mothmen.dm [hunk 5]
```diff
@@ -28,7 +28,7 @@
 	payday_modifier = 0.75
 	family_heirlooms = list(/obj/item/flashlight/lantern/heirloom_moth)
 
-/datum/species/moth/regenerate_organs(mob/living/carbon/C,datum/species/old_species,replace_current=TRUE,list/excluded_zones)
+/datum/species/moth/regenerate_organs(mob/living/carbon/C,datum/species/old_species,replace_current=TRUE,list/excluded_zones,visual_only)
```
Comment (SuperNovaa41) [inline code/modules/mob/living/carbon/human/species_types/mothmen.dm:31]:
```suggestion
/datum/species/moth/regenerate_organs(mob/living/carbon/C, datum/species/old_species, replace_current= TRUE, list/excluded_zones, visual_only)
```
i know you only added the last part but wouldn't it be quirky if you updated the whoel thing

**Review by SuperNovaa41 — COMMENTED** [review]
i wanted to rip this thing apart but you actually follow the damn style guide

**Comment by Mothblocks** [comment]
@LemonInTheDark Fix the conflict and I'll TM this on Sybil

**Comment by LemonInTheDark** [comment]
I **guess** I could do that

> i wanted to rip this thing apart but you actually follow the damn style guide

Oh, you think darkness is your ally. But you merely adopted the dark; I was born in it, moulded by it. I didn't see the light until I was already a man, by then it was nothing to me but BLINDING!

File: code/modules/mob/living/carbon/human/dummy.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/carbon/human/dummy.dm:31]: sure, thx

File: code/modules/mob/living/carbon/human/species_types/mothmen.dm (hunk 5 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/carbon/human/species_types/mothmen.dm:31]: It's a bit long, but sure I'm good with that, not long enough to be a problem

File: code/modules/mob/living/carbon/human/dummy.dm [hunk 6]
```diff
@@ -57,20 +57,23 @@ INITIALIZE_IMMEDIATE(/mob/living/carbon/human/dummy)
 			to_nuke += checking
 			continue
 
+		if(istype(checking, /obj/item/pda))
+			var/obj/item/pda/whats_shitcode = checking
```
Comment (Mothblocks) [inline code/modules/mob/living/carbon/human/dummy.dm:61]: ahem

Comment (Mothblocks) [reply code/modules/mob/living/carbon/human/dummy.dm:61]: damn this isnt even yours this sucks

**Comment by Mothblocks** [comment]
@LemonInTheDark Please explain what the bug was that you fixed and how you fixed it, so that I can understand the kind of quirks this system will introduce

File: code/modules/mob/living/carbon/human/dummy.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/carbon/human/dummy.dm:61]: (he doesn't know)

**Comment by LemonInTheDark** [comment]
> @LemonInTheDark Please explain what the bug was that you fixed and how you fixed it, so that I can understand the kind of quirks this system will introduce

Sure!
There were two sources of failure.

1: Rainbow jumpsuits
The way I had provide_type setup, if you passed in a location no matter what the target would just be force moved to it, if we just new'd it, or if it was old.

Rainbow jumpsuit temporarys del on new, so we were spawning them in deleted. Not good.

2: PDA gear

PDAs hold refs to their cartridge and pen. I'd like to cache those things, but it requires pulling them out of the pda before recaching them.

I pulled then out just fine, I just did it after a check that caught the pda. So it didn't really work so well.

File: code/modules/mob/living/carbon/human/dummy.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/carbon/human/dummy.dm:61]: I should put this on move logic, if you remove a pen from the pda it should lose the ref no matter what

**Comment by Mothblocks** [comment]
Is #1 possible to unit test?

Was #2 just a hard-del or was it causing other big issues?

File: code/modules/mob/living/carbon/human/dummy.dm (hunk 6 above)
Comment (Mothblocks) [reply code/modules/mob/living/carbon/human/dummy.dm:61]: its just a bad name

**Comment by LemonInTheDark** [comment]
#1 was a flaw in the underlying system, unit test caught it actually
#2 was a hard del -> nulls in the object store -> people get "spawned" no gear. I'm going to add logic to prevent that

File: code/modules/mob/living/carbon/human/dummy.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/carbon/human/dummy.dm:61]: no like outside of the name this is shitty code

**Comment by Mothblocks** [comment]
Cool, I only bring all this up so that you find ways to prevent stuff like this happening in the future. Basically what we talked about before--bugs involving this system are going to be extremely weird.

Ready for another TM?

**Comment by LemonInTheDark** [comment]
Lemme fix the unit test failure first

**Comment by LemonInTheDark** [comment]
I wonder why the bot only now noticed

**Comment by optimumtact** [comment]
for clarities sake this breaks spawning of plasmamen tanks in survival boxes as it assumes the target mob will be a plasmaman, which the cached version will not take into account.

**Comment by LemonInTheDark** [comment]
I've fixed the box error, @Mothblocks this is one of the weird ones, not a good way to test for it.
Unfucked the pda code too, happy with this

File: code/modules/mob/living/carbon/human/dummy.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/carbon/human/dummy.dm]: Shit is unfuck

**Comment by LemonInTheDark** [comment]
Time is a flat circle https://github.com/tgstation/tgstation/pull/23150

This one's better though, because it's not trying to be general, we skip Initialize and Destroy. Actually saves cpu time
